### PR TITLE
[DisplayList] migrate DlImageFilter code to Impeller geometry classes

### PR DESCRIPTION
**Pull request by flar**
The DlImageFilter code uses Skia geometry classes for its internal computations. This PR switches those implementations to use the Impeller geometry classes for consistency and 3rd party header file independence.

### CODE REVIEW

**Comment by flar** [comment]
There are just a few lines of code in dl_image_filter.cc that still use the Skia classes for bounds computation as they seem to pass the perspective transform tests better, but otherwise the ImageFilter code is 99.5% migrated to the Impeller classes including what should be 100% of the API (not including the use of SkShader by the runtime effect class). See https://github.com/flutter/flutter/issues/159175 for more information.

**Comment by flar** [comment]
I find the golden diffs interesting. They are very tiny, but they don't appear (by name) to involve image filters.

File: display_list/dl_builder.cc [hunk 1]
```diff
@@ -437,7 +438,7 @@ void DisplayListBuilder::saveLayer(const DlRect& bounds,
 
   // A backdrop will affect up to the entire surface, bounded by the clip
   bool will_be_unbounded = (backdrop != nullptr);
-  std::shared_ptr<const DlImageFilter> filter;
+  std::shared_ptr<DlImageFilter> filter;
```
Comment (gaaclarke) [inline display_list/dl_builder.cc:441]: Why change this?

File: display_list/dl_builder.cc [hunk 2]
```diff
@@ -717,15 +719,17 @@ void DisplayListBuilder::TransferLayerBounds(const SkRect& content_bounds) {
...
+    if (!global_bounds.IsEmpty()) {
+      DlIRect global_ibounds = DlIRect::RoundOut(global_bounds);
       if (!filter->map_device_bounds(global_ibounds, matrix, global_ibounds)) {
         parent_is_flooded = true;
       } else {
-        global_bounds.set(global_ibounds);
-        if (global_bounds.intersect(clip)) {
-          parent_layer().global_space_accumulator.accumulate(global_bounds);
+        global_bounds = DlRect::Make(global_ibounds);
+        auto clipped_bounds = global_bounds.Intersection(clip);
```
Comment (gaaclarke) [inline display_list/dl_builder.cc:729]: https://google.github.io/styleguide/cppguide.html#Type_deduction

File: display_list/dl_builder.h [hunk 3]
```diff
@@ -52,7 +56,7 @@ class DisplayListBuilder final : public virtual DlCanvas,
   void Save() override;
 
   // |DlCanvas|
-  void SaveLayer(std::optional<const DlRect>& bounds,
+  void SaveLayer(const std::optional<DlRect>& bounds,
```
Comment (gaaclarke) [inline display_list/dl_builder.h:59]: Why was this changed?

File: flow/layers/image_filter_layer.cc [hunk 4]
```diff
@@ -10,12 +10,12 @@
 
 namespace flutter {
 
-ImageFilterLayer::ImageFilterLayer(std::shared_ptr<const DlImageFilter> filter,
+ImageFilterLayer::ImageFilterLayer(const std::shared_ptr<DlImageFilter>& filter,
```
Comment (gaaclarke) [inline flow/layers/image_filter_layer.cc:13]: I don't think switching these to const references is an improvement.  With the std::move there ways to avoid a copy, the const reference forces a copy to happen.

File: flow/layers/layer_state_stack.h [hunk 5]
```diff
@@ -278,6 +278,12 @@ class LayerStateStack {
   // This rectangle may be a conservative estimate of the true clip region.
   SkRect local_cull_rect() const { return delegate_->local_cull_rect(); }
 
+  // The transform from the local coordinates to the device coordinates
+  // in 4x4 DlMatrix representation. This matrix provides all information
+  // needed to compute bounds for a 2D rendering primitive, and it will
+  // accurately concatenate with other 4x4 matrices without losing information.
```
Comment (gaaclarke) [inline flow/layers/layer_state_stack.h:284]: Please write as a docstring (ie `///`)

File: impeller/geometry/rect.h [hunk 6]
```diff
@@ -151,6 +151,14 @@ struct TRect {
     return TRect(0.0, 0.0, size.width, size.height);
   }
 
+  template <class U, class FT = T>
```
Comment (gaaclarke) [inline impeller/geometry/rect.h:154]: docstring would help here, it's not immediately clear to me what this is for.  it's copying a rect, but why the `enable_if_t`?  If this potentially throwing away precision implicitly?

**Review by gaaclarke — COMMENTED** [review]
Woo, that was a lot to look at.  Mostly looking good, here's some notes and questions.

File: impeller/geometry/matrix.h [hunk 7]
```diff
@@ -305,6 +305,8 @@ struct Matrix {
 
   Scalar GetDeterminant() const;
 
+  bool IsInvertible() const { return GetDeterminant() != 0; }
```
Comment (jonahwilliams) [inline impeller/geometry/matrix.h:308]: can you put these in the implementation?

File: display_list/dl_builder.cc (hunk 1 above)
Comment (flar) [reply display_list/dl_builder.cc:441]: The classes are immutable and I was having trouble with the fact that you can't mix and match shared_ptr with different "const" modifiers.

File: display_list/dl_builder.h (hunk 3 above)
Comment (flar) [reply display_list/dl_builder.h:59]: const in an optional doesn't really prevent you from changing it, the optional itself has to be const. And at that point, why have the const inside the template type?

File: flow/layers/image_filter_layer.cc (hunk 4 above)
Comment (flar) [reply flow/layers/image_filter_layer.cc:13]: These are always called from a reference source that cannot be moved so a copy must happen at some point. It would be better if the copy happened into the layer's local field rather than to the value handed in to the constructor.

File: flow/layers/layer_state_stack.h (hunk 5 above)
Comment (flar) [reply flow/layers/layer_state_stack.h:284]: None of the methods in this file are docstring. This is an issue with a scope outside of this change - https://github.com/flutter/flutter/issues/159179

File: impeller/geometry/rect.h (hunk 6 above)
Comment (flar) [reply impeller/geometry/rect.h:154]: Done. It's only defined for float destinations. If you want to convert to an integer rect you have to use a method like RoundOut to specify the rounding behavior.

File: impeller/geometry/matrix.h (hunk 7 above)
Comment (flar) [reply impeller/geometry/matrix.h:308]: I'm not sure what you are asking?

File: display_list/dl_builder.cc (hunk 2 above)
Comment (flar) [reply display_list/dl_builder.cc:729]: Done

File: impeller/geometry/matrix.h (hunk 7 above)
Comment (jonahwilliams) [reply impeller/geometry/matrix.h:308]: I forget the proper terminology :). Can you put the method bodies in the .cc file?

Comment (flar) [reply impeller/geometry/matrix.h:308]: These geometry files seem to prefer inline implementations unless the method is very complicated.

**Review by jonahwilliams — APPROVED** [review]
LGTM!

File: display_list/dl_builder.h (hunk 3 above)
Comment (gaaclarke) [reply display_list/dl_builder.h:59]: > why have the const inside the template type?

To make sure no one modifies the `DlRect`.  I would have expected `const std::optional<const DlRect>& bounds`

File: flow/layers/image_filter_layer.cc (hunk 4 above)
Comment (gaaclarke) [reply flow/layers/image_filter_layer.cc:13]: You are considering just the code that is written today.  There is no way to enforce that this is only called with references.

File: display_list/dl_builder.cc (hunk 1 above)
Comment (gaaclarke) [reply display_list/dl_builder.cc:441]: https://google.github.io/styleguide/cppguide.html#Use_of_const

Immutable classes should have const methods so having a const reference should still be the appropriate usage in c++.
